### DISCUSSION TITLE
Using libjupitermag in a multi-threaded program

### DISCUSSION THREAD

**Discussion by cpiker** [mattkjames7/libjupitermag#5] [General]
I'm tasked with creating a data stream output program that produces M-Shell values using JRM33 and con2022 at Juno's position.  Since field line tracing without an analytical solution can take a bit of time, are the various trace routines multi-thread safe after initial configuration?

Also, what are the best functions to use for getting Juno's M-Shell using libjupitermag?  

Thanks a lot for creating this library by the way.  The roll-up C and C++ compatible header is working rather well.

**Comment by mattkjames7** [comment]
Thanks for showing interest in our code!

Good question about the code being thread-safe. I suspect that it _probably_ is as long as the model configuration doesn't change from one trace to another, though we may need to dig into it a bit. 

I'm no Jupiter expert, but if M-shell is what I think it is, and you're using the C wrapper function for the trace, then I would say that element 46 of the `FP` array might be what you're looking for (on line 641 of [the header](https://github.com/mattkjames7/libjupitermag/blob/main/include/jupitermag.h) ). That is the radial distance of the field line as it crosses the magnetic equator. I should really put a bit more documentation in the code!

I have asked my colleagues about the M-shell to be sure about my answer!

**Comment by cpiker** [comment]
I spoke a bit out of turn asking about threading.  I hadn't run a test case yet so my run-time expectations were based on similar IDL code.  So far the performance of libjupitermag Trace routines is amazing, (~23 ms/trace out at 37 Rj on a core i5 @ 2.5 GHz) so I'm not that worried about threading any more.  Nice job!

Just to make sure I'm using the library properly.  When I invoke the `TraceField()`  C wrapper and provide `jrm33` and `con2020` as the internal and external fields, am I'm I getting the hybrid solution to con2020 as described in section 4.2 [the SSR 2022 paper](https://doi.org/10.1007/s11214-023-00961-3) ? 

I'm concerned that I'm only invoking the analytic model and so `TraceField` run-times don't include integration periods.  I'm using the following to set the input positions...
```d
x0 = 1.2; y = 0.0; z0 = 0.0;
foreach(i; 0..30){
   x = x0 + i*0.2;
   z = z0 + i*0.2;

   TraceField(1, &x, &y, &z, "jrm33", 1, &"con2020", ...);
}
```
...but the code seems to run too fast to be using Connerney et al. integrals.

**Comment by mattkjames7** [comment]
The default settings should be using the hybrid model. One way to be certain would be to call the `GetCon2020Params` function to retrieve the current parameters. 23ms does seem quite quick, but I suppose that  the number of steps which require the integral solution depends on where the trace originates and the step size. I can say for certain that using the purely integral solution for entire traces is _slow_ - a few seconds per trace from memory.

**Comment by cpiker** [comment]
I'm calling GetCon2020Params and integrating the output into stream metadata.  Client tool says the model is running in hybrid mode:

![model_metadata](https://github.com/mattkjames7/libjupitermag/assets/7530814/f6a70f9b-bd02-4c7b-97a2-f932f3a1b6d1)

but it's still running quite fast so maybe I'm just not hitting the inner edge of the current sheet where integration would be needed.  I'll try earlier Juno orbits and see.

**Reply by mattkjames7** [reply]
I had a quick look in the code and I think that output you have there looks how I expect it to. From memory, I think the C++ code should be capable of ~20 vectors/s using the purely integral model, to tracing can take a bit of time, but individual vectors should be fairly quick still.
